### DISCUSSION TITLE
[gltf-transform/cli] Missing required extension, "MSFT_texture_dds"

### DISCUSSION THREAD

**Discussion by akukh** [donmccurdy/glTF-Transform#1444] [Q&A] [answered]
Hi,

I need to merge several different glTF files into one (preferably into one scene). I found your CLI library with a command to be able to merge multiple files, which seemed like the perfect solution for me. However, when I tried to merge, I got the error:
```Shell
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".
warn: Missing optional extension, "MSFT_texture_dds".

error: Missing required extension, "MSFT_texture_dds".
```

Is there any way to fix/enable support for this? The option to simply ignore this extension could also be suitable.

P.S. I tried the Blender option, but to open multiple glTF files and then merge them -- consumes an abnormal amount of time and RAM (and not always successful).

**Comment by donmccurdy** [comment]
To add support for an extension to the CLI, you'll need to use a [CLI configuration file](https://gltf-transform.dev/cli-configuration). If you're OK with ignoring the extension, that could just contain a simple stub of the extension like this:

```javascript
// gltf-transform.config.mjs
import { Extension } from '@gltf-transform/core';
import { ALL_EXTENSIONS } from '@gltf-transform/extensions';

// NOTE: Stub implementation.
class MSFTTextureExtension extends Extension {
 static EXTENSION_NAME = 'MSFT_texture_dds';
 extensionName = 'MSFT_texture_dds';
 write(context) { return this; }
 read(context) { return this; }
}

export default {
 extensions: [...ALL_EXTENSIONS, MSFTTextureExtension]
};
```

```bash
gltf-transform merge a.glb b.glb c.glb out.glb --config path/to/gltf-transform.config.mjs
```

You may find it easier to do this in a script than through the CLI — if so, see [`mergeDocuments`](https://gltf-transform.dev/modules/functions/functions/mergeDocuments).

**Reply by akukh** [reply]
Thanks for the answer. I tried the config option for the CLI, but I get the following errors:
```Shell
> gltf-transform --help --config .\merge-ignore-ext.config.mjs

error: Cannot find package '@gltf-transform/core' imported from D:\Scenes\05-06-2024\merge-ignore-ext.config.mjs

node:internal/modules/esm/resolve:854
  throw new ERR_MODULE_NOT_FOUND(packageName, fileURLToPath(base), null);
        ^

Error [ERR_MODULE_NOT_FOUND]: Cannot find package '@gltf-transform/core' imported from D:\Scenes\05-06-2024\merge-ignore-ext.config.mjs
    at packageResolve (node:internal/modules/esm/resolve:854:9)
    at moduleResolve (node:internal/modules/esm/resolve:927:18)
    at defaultResolve (node:internal/modules/esm/resolve:1157:11)
    at ModuleLoader.defaultResolve (node:internal/modules/esm/loader:383:12)
    at ModuleLoader.resolve (node:internal/modules/esm/loader:352:25)
    at ModuleLoader.getModuleJob (node:internal/modules/esm/loader:227:38)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/module_job:87:39)
    at link (node:internal/modules/esm/module_job:86:36) {
  code: 'ERR_MODULE_NOT_FOUND'
}
```
What am I doing wrong? Does the name of the config file matter, or its path (where it is located)? I definitely installed via npm `@gltf-transform/core` and `@gltf-transform/extensions`. I'm using Windows, if that can be of any help. Furthermore, I'm sorry if these are dumb questions, but I'm not very familiar with nodejs or anything JavaScript related😅

**Reply by donmccurdy** [reply]
Not a dumb question at all! The "CLI config" system is experimental at this point, I think most people go either pure CLI (without a config) or set up scripting if they need more control or custom extensions. There might be other ways to handle this (like global installs) but I think you'd need the config file to be within the scope of an npm workspace with these extensions installed. Something like:

```bash
# cwd: path/to/my-project
npm install --save @gltf-transform/core @gltf-transform/extensions
gltf-transform merge a.glb b.glb out.glb --config ./gltf-transform.config.mjs
```

**Reply by akukh** [reply]
Thank you again, everything worked. But am I correct in my understanding that I can write my own script (at least in theory) that will be able to merge `glTF` files with the extension `MSFT_texture_dds`? Using only the `gltf-transform` library.

**Reply by donmccurdy** [reply, marked as answer]
Yes, that's quite possible. To retain the `MSFT_texture_dds` extension in the merged files, you'll need to fully implement it. See the implementation of `KHR_texture_basisu` which should be nearly identical:

https://github.com/donmccurdy/glTF-Transform/blob/08c55732b19df75ffce832278ca310568cc6e96d/packages/extensions/src/khr-texture-basisu/texture-basisu.ts#L124-L174

If you only need to merge documents, and not to edit them more deeply, then I believe the `ImageUtils.registerFormat` and ImageUtils-related code can be skipped. See [`merge.ts`](https://github.com/donmccurdy/glTF-Transform/blob/main/packages/cli/src/transforms/merge.ts) for an example of how the merging could be implemented, but depending on your goals some parts of that could be skipped too.
